modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/QueryMemoryTracker.java

```diff
@@ -203,74 +235,239 @@ public boolean isOffloadingEnabled() {
-                ", bytesOffloaded=" + totalWrittenOnDisk + ", filesCreated=" + filesCreated +
-                ", query=" + qryDesc + ']');
-        }
+            parent.release(reservedFromParent);
     }
 
-    /**
-     * @return Total number of bytes written on disk.
-     */
-    public long totalWrittenOnDisk() {
-        return totalWrittenOnDisk;
+    /** {@inheritDoc} */
+    @Override public synchronized void incrementFilesCreated() {
+        if (parent != null)
+            parent.incrementFilesCreated();
+
+        filesCreated++;
     }
 
     /** {@inheritDoc} */
-    @Override public synchronized void addTotalWrittenOnDisk(long written) {
-        this.totalWrittenOnDisk += written;
-    }
+    @Override public synchronized H2MemoryTracker createChildTracker() {
+        checkClosed();
 
-    /**
-     * @return Total bytes reserved by current query.
-     */
-    public long totalReserved() {
-        return totalReserved;
-    }
+        H2MemoryTracker child = new ChildMemoryTracker(this);
 
-    /**
-     * @return Total files number created by current query.
-     */
-    public int filesCreated() {
-        return filesCreated;
-    }
+        children.add(child);
 
-    /** {@inheritDoc} */
-    @Override public synchronized void incrementFilesCreated() {
-        this.filesCreated++;
+        return child;
     }
 
-    /**
-     * @return Query descriptor.
-     */
-    public String queryDescriptor() {
-        return qryDesc;
+    /** {@inheritDoc} */
+    @Override public synchronized void onChildClosed(H2MemoryTracker child) {
+        if (state != CLOSING_OR_CLOSED_STATE)
+            children.remove(child);
     }
 
     /** {@inheritDoc} */
     @Override public String toString() {
         return S.toString(QueryMemoryTracker.class, this);
     }
-}
\ No newline at end of file
+
+    /** */
+    private static class ChildMemoryTracker implements H2MemoryTracker {
+        /** State updater. */
+        private static final AtomicIntegerFieldUpdater<ChildMemoryTracker> STATE_UPDATER
+            = AtomicIntegerFieldUpdater.newUpdater(ChildMemoryTracker.class, "state");
+
+        /** */
+        private final H2MemoryTracker parent;
+
+        /** */
+        private long reserved;
+
+        /** */
+        private long writtenOnDisk;
+
+        /** */
+        private long totalWrittenOnDisk;
+
+        /** */
+        private volatile int state;
+
+        /**
+         * @param parent Parent.
+         */
+        public ChildMemoryTracker(H2MemoryTracker parent) {
+            this.parent = parent;
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean reserve(long size) {
+            checkClosed();
+
+            boolean res;
+            try {
+                res = parent.reserve(size);
+            }
+            catch (IgniteSQLException ex) {
+                reserved += size;
```

Comment:
I don't understand that logic _reserve anyway even if parent reservation fails_. 
Why don't use `finally` here?

Reply:
Yes, `finally` is fit better, thanks! 
We should count memory here to match tracked memory on parent (parent reserves memory, but then could throw exception and doesn't release memory)